PHR IPv4

scgruber opened the pull request:
This is a working branch for the IPv4 implementation. Will be splitting the commits into more logical pieces and adding tests before this is okay to merge in.

At this point I think the structure of the link is complete though.

scgruber committed:
PHR IPv4: Send traffic for the WAN through NAT

CollinValley commented on examples/pi-home-router/src/interface/link/interface_collect.rs:
```diff
@@ -89,7 +89,7 @@ mod tests {
         let host = immediate_stream(packets.clone());
         let lan = immediate_stream(packets.clone());
-        let wan = immediate_stream(packets.clone());
+        let wan = immediate_stream(packets);
```
Good catch

scgruber replied:
I can only praise our lord and savior clippy for this one.

CollinValley commented on examples/pi-home-router/src/interface/link/interface_dispatch.rs:
```diff
@@ -71,17 +71,17 @@ impl LinkBuilder<InterfaceAnnotated<EthernetFrame>, EthernetFrame> for Interface
         let (_, host_e) = ProcessLink::new()
             .ingressor(ce.remove(0))
-            .processor(InterfaceAnnotationDecap)
+            .processor(InterfaceAnnotationDecap::new())
```
Unfortunate that we need the PhantomData with generic structs. But it makes sense.

CollinValley commented on examples/pi-home-router/src/interface/processor/annotation_set_out.rs:
```diff
@@ -0,0 +1,35 @@
+use route_rs_runtime::processor::Processor;
+use std::marker::PhantomData;
+
+/// InterfaceAnnotationEncap: Processor to apply interface annotation to a packet
```
Need to update the doc comment here.

CollinValley replied:
Consider also using a different shortening for output.  I was a bit confused by the name when I saw it.

CollinValley commented on examples/pi-home-router/src/ipv4/mod.rs:
```diff
@@ -0,0 +1,388 @@
+
+#[allow(dead_code)]
+impl HandleIPv4 {
+    pub(crate) fn new() -> Self {
```
Thinking about how configuration will work in our router, from the top level. Perhaps we will need to create some sort of configuration structure.  Think how C libs commonly take a bunch of arguments, you make a structure that gets passed in, that struct is used to set opts.  Perhaps we could also just accept a JSON object as an argument or something?  Nothing to worry about for now, something to consider if the number of config options gets high.

scgruber replied:
Yeah I think figuring out the right balance between explicitness and ease of passing configuration is going to be pretty important, and I'm not sure exactly what to do about that yet.

CollinValley commented on examples/pi-home-router/src/interface/processor/annotation_decap.rs:
```diff
@@ -1,14 +1,24 @@
 use crate::types::InterfaceAnnotated;
-use route_rs_packets::EthernetFrame;
+use route_rs_packets::Packet;
```
Thanks for expanding this to be generic for me.

CollinValley commented on examples/pi-home-router/src/interface/processor/annotation_encap.rs:
```diff
@@ -1,28 +1,31 @@
 use crate::types::{Interface, InterfaceAnnotated};
-use route_rs_packets::EthernetFrame;
+use route_rs_packets::Packet;
```
Same as above: 👍

CollinValley commented on examples/pi-home-router/src/ipv4/mod.rs:
```diff
@@ -0,0 +1,388 @@
+        }
+    }
+
+    fn build_link(self) -> Link<InterfaceAnnotated<Ipv4Packet>> {
```
You should consider breaking this into more than just one mega-link.  Ideally each build_link, like a function, doesn't get this long.  I know this is a draft so it is fine right now.

scgruber replied:
I don't think this is actually that complex of a logic, there are maybe 20 links in here. The problem I think is that setting up each element is this huge block even though conceptually there's not that much going on.

Personally I don't like trying to split things into so many pieces just for the goal of splitting them up, because it can become much harder to follow the flow of what's happening.

This is also an area where I expect the graph view to be much easier to interact with. I sketched a graph on paper when I was building this and it was pretty straightforward to follow, even though linearizing it in code is less clear.

CollinValley commented on examples/pi-home-router/src/ipv4/mod.rs:
```diff
@@ -0,0 +1,388 @@
+
+        // Since all NAT encapsulated traffic is from the LAN bound for the WAN, we can just
+        // regenerate the interface annotations here. Traffic from the Host is marked as LAN here,
+        // but that's not really inaccurate and we won't use this information later anyway.
```
Perhaps just leave the Inbound_inteface as unmarked then? To convey to the downstream that it is unknown information as this point.  You could also probably rebuild it (host vs LAN) just by checking the src ip of the packet.

CollinValley replied:
Ideally we don't ever add incorrect information to a packet.

CollinValley commented on examples/pi-home-router/src/ipv4/mod.rs:
```diff
@@ -0,0 +1,388 @@
+
+        // Send traffic from the WAN to the NAT decapsulator.
+        //
+        // TODO: Implement NAT decapsulator
```
Decapsulator is a good name, sounds like an evil villain weapon.

CollinValley reviewed (commented):
Overall, this is really exciting, awesome work, thanks for your help!

scgruber committed:
PHR: Make annotation processors generic

These processors should work over generic Packet structs, not just
EthernetFrame. This will allow us to promote and demote packets
between the different protocols in the router.

scgruber committed:
PHR: HandleIpv4 trivial link

Set up the basic link structure, but don't actually transform anything.

scgruber committed:
PHR: Classify packets based on inbound interface

This will allow us to make routing decisions based on whether packets
are coming from the lan or wan. Since the types allow us to receive
packets with an unmarked inbound interface, we drop them because we
don't know what to do. Include a test to document that this is intended.

ddowl commented on examples/pi-home-router/src/interface/classifier/inbound_interface.rs:
```diff
@@ -0,0 +1,30 @@
+    type Class = Interface;
+
+    fn classify(&self, packet: &Self::Packet) -> Self::Class {
+        match packet.inbound_interface {
```
just return `packet.inbound_interface`.





bruh

ddowl commented on examples/pi-home-router/src/ipv4/mod.rs:
```diff
@@ -0,0 +1,473 @@
+            classify_by_inbound_interface_egressors.remove(0);
+
+        // Packets from the Host either are bound for the LAN subnet (or LAN broadcast), else they
+        // go out to the WAN. Broadcast packets shouldn't be sent up to the LAN because it is a
```
oops this should be "sent up to the WAN"

ddowl commented on examples/pi-home-router/src/ipv4/classify_dest_subnet.rs:
```diff
@@ -0,0 +1,45 @@
+/// This classifier uses a HashMap of CIDRs to subnet names. We require but do not enforce that the
+/// CIDRs are disjoint. Undefined behavior will occur if they are overlapping.
+///
+/// TODO: Actually enforce that CIDRs are disjoint at runtime.
```
Actually let's allow overlaps but prefer the longest CIDR to match. We can sort the keys by length once in the constructor for performance.

scgruber committed:
PHR IPV4: Drop spurious packets from the LAN

Packets from the LAN should be destined for the Host (or broadcast) or
the WAN. Any packets that we receive destined for other LAN addresses
are an error, and we should drop them.

In order to implement this, we now have to know what our LAN IP and
subnet are, so we add fields and builders for those.

Note that we have to handle both the global broadcast 255.255.255.255
and the local broadcast which is (network | (broadcast & subnet mask)).
Fortunately, cidr provides .last_address() which does this for us.

scgruber committed:
runtime: unpack_link! macro

This macro is a different syntax for unpacking to be used after the
link is built. This has a few advantages:
- IntelliJ code inspection properly highlights and links the code in
  the LinkBuilder block, which is basically all of the meaningful code
  in a link defintion.
- cargo fmt and cargo clippy actually format and inspect the code in
  the LinkBuilder block.
- link => runnables_collector, [egressor, ...] feels like a more
  accurate syntax than a fake let, since the runnables are appended,
  not assigned.
- It allows us to give a name to the link which is probably good for
  readability of big composites.

scgruber committed:
PHR IPV4: Tag packets from lan to host

Implement InterfaceAnnotationSetOutbound to mark the outbound interface
on a packet. We use this right now to mark packets from the lan to the
host, as classified previously.

scgruber committed:
PHR IPV4: Classify packets from Host

Additionally, if we know they go out the LAN, we can simply tag them
for that at this point. WAN packets require NAT logic that will be
filled in by a later commit.

scgruber committed:
PHR IPV4: Route packets from NAT for Host

If something comes out of the NAT with our destination IP address, we
need to send it to the Host interface, not the LAN. This functionality
is added, but we can't test it yet because it requires the NAT sublink.

scgruber committed:
PHR IPV4: Set correct input interfaces after NAT

After NAT, we regenerate the input interface annotation because it was
removed. For packets from the host, we need an extra bit of logic to
set their input interface based on the source subnet.

scgruber committed:
Update cargo-deny to 0.6.6

One of the cargo-deny dependencies started failing, probably because
the compiler has been bumped a lot since we last updated. Turns out
cargo-deny has also changed a lot, so this commit includes some
updates to our config. It doesn't change the allowlist.

CollinValley commented on route-rs-runtime/src/lib.rs:
```diff
@@ -65,3 +65,34 @@ macro_rules! unpack {
 }
+
+#[macro_export]
+macro_rules! unpack_link {
```
This looks good, I like the benefits of this newer layout.  However can you make this one supplant the old one?  Choose either name.

scgruber replied:
Yeah I just didn't want to go through all the trouble of converting everywhere before someone else approved it. Will update things now.

CollinValley commented on examples/pi-home-router/src/interface/processor/annotation_set_inbound.rs:
```diff
@@ -0,0 +1,34 @@
+use route_rs_runtime::processor::Processor;
+use std::marker::PhantomData;
+
+/// InterfaceAnnotationSetOutbound: Update the outbound interface for a packet
```
Doc comment needs updating.

CollinValley reviewed (changes requested):
Just a few final things, then I think this is ready for merge.  I really like the trouble you've went to to provide good test coverage as well, makes the code a lot more professional.

scgruber commented:
Should be good to go now
